### ISSUE TITLE
getColums() do not reflect actual column ordering if reordering allowed

### ISSUE THREAD

**Issue by martinisraelsen** [vaadin/vaadin-grid-flow#338]
Grid.getColumns() returns the original column ordering, not the actual column ordering that is shown in the grid. 

This is problematic because an application should be able to remember the users configuration (column order and column width) and reuse those settings next time the grid is shown. As it is now, the grid will be reset to the original settings.

**Comment by jouni** [comment]
As an implementation note: the “physical” (i.e. DOM) order of the columns does not change in the browser. The change is only visual, done with CSS.

I’m wondering if a “column reorder” event would suffice as the first step? Though, I don’t mind if the Flow integration wants to do the mapping internally if that is the expected behavior from developers.

**Comment by Klaudeta** [comment]
If the column reorder event with only hold information on previous and new position of the column, I think it will be required also for Flow component to keep track of the actual order of the columns and same time use the column reorder event to update that order

Is there any plan to have this the column reorder event available?

**Comment by jouni** [comment]
> I think it will be required also for Flow component to keep track of the actual order of the columns

Well, not necessarily. Flow can also just expose the event for the developer, and then the developer takes care of keeping track of the visual order.

> Is there any plan to have this the column reorder event available?

Not that I know of. I don’t think there’s an issue for that in the vaadin-grid repo.

**Comment by Klaudeta** [comment]
It looks a bit odd to me to have the user store and manage the order of the columns outside the grid when common sense make you think that getColumns() should return also the exact order. This is how I see it at least.

And for the column reorder event I just found this https://github.com/vaadin/vaadin-grid/pull/1568 that is already merged.

**Comment by jouni** [comment]
> And for the column reorder event I just found this vaadin/vaadin-grid#1568 that is already merged.

Great! Does the Flow Grid expose that (I assume it does)? Then the workaround is there already for @martinisraelsen, to keep track of the order.

But I agree, it is cumbersome, and Flow Grid could help with that.

**Comment by Klaudeta** [comment]
No it doesn't (yet) and I think this is the first step in here to do

**Comment by pleku** [comment]
Sorry for the mixup, but since the javadocs in the [`getColumns` method clearly state](https://vaadin.com/api/platform/13.0.1/com/vaadin/flow/component/grid/Grid.html#getColumns--) that the ordering is not guaranteed to match the client side, this is not a bug but an enhancement request. 

Luckily the support for this has already been added to the web component, so it can be now implemented for the component Java integration too. This should go to the next minor release of the component Java integration that uses the new version of the web component.

**Comment by spikeon** [comment]
Has there been any movement on this getting into flow?  We need to store the user's preferred UI configuration between sessions, and need to get re-ordering and store it in the database. 

Obviously, a listener would be ideal, but since that doesn't exist, I was hoping to get the columns in whatever order they are in using flow in a thread, and am therefore waiting on this issue.

**Comment by petrixh** [comment]
Just adding a comment and sub here as this is relevant for a customer project that I'm working on. Also relevant is an API for listening to the reordering events, but maybe that's given.

**Comment by mvysny** [comment]
Unfortunately this is tricky to implement. I've created a DOM event class in order to listen for the `column-reorder` event:
```kotlin
@DomEvent("column-reorder")
class ColumnReorderEvent<T>(source: Grid<T>, fromClient: Boolean, @EventData("event.detail.columns") val columns: List<Grid.Column<T>>) : ComponentEvent<Grid<T>>(source, fromClient)
```
However unfortunately Flow doesn't currently support transferring references to elements to the server-side: https://github.com/vaadin/flow/issues/6372

A workaround would be to transfer only the column String IDs (an idea is to use `@EventData("event.detail.columns.map(col -> col.id)") val columns: List<String>`) however unfortunately the `col.id` is always an empty string.

**Comment by tomivirkki** [comment]
@mvysny If possible, align with https://github.com/vaadin/vaadin-grid-flow/pull/746/files#diff-18fd66d51777465fec3d9be42bac947cR915

**Comment by mvysny** [comment]
Thank you @tomivirkki the `_flowId` property worked well and in fact returns the correct column ID (e.g. `"col0"`).

We have agreed to not to modify the `Grid.getColumns()` method as of now - so it will continue to not to reflect client-side ordering. So strictly speaking, we can close this issue as WON'T FIX.

I will however open a new feature request to implement support for the `column-reorder` feature, and work on that.
